### ISSUE TITLE
Creating a Bias layer for use with Functional API

### ISSUE THREAD

**Issue by quantlearn** [fchollet/keras#5054]
I am building a model using the Functional API and require a custom layer for a bias term.   I have attempted to implement a custom layer called Bias which outputs a vector of length output_dim.  

```
class Bias(Layer):
 ...
    def call(self, x=None,mask=None):
        output = self.b
        return self.activation(output)

    def get_output_shape_for(self, input_shape):
            return (1, self.output_dim)


```

Usage:

```
bias = Bias(N)()
merged = merge([X,bias],mode='concat')
```

The problem is that the call function expects an input tensor even if as in this case the input tensor is not used.  I have attempted to produce a work around but keep getting various Keras errors.   

Please can someone suggest an appropriate solution for the Bias layer.

Many thanks.

**Comment by bstriner** [comment]
You need an input tensor because that is how you know the batch dimension. Even if you don't use the contents of the input tensor, you will need to use `K.shape(x)[0]` to get the batch dimension. You could then do `K.repeat_elements` to repeat the bias to be the right shape.

Let's say your model only needs targets, not inputs, you will still pass in a dummy input `np.ones(...)` that is the right batch dimension.

Another possible solution is to just use a Dense layer, but set the weight matrix to 0 and remove it from trainable_weights, so W is always 0 and it only learns b.

Cheers,
Ben

**Comment by quantlearn** [comment]
Ben,

Thank you for your reply.   I have a working implementation but is not as elegant as I would like.   

There is actually no need to have `W` as the implementation of `call` only requires a `b`.  It is also possible to explicitly pass the `batch_size` as this is known.   The issue is that Keras has a number of checks in `topology.py` that require as input Tensor even if it is not used.   I would like to understand how to write a custom layer that does not need to put unnecessary nodes in the compute graph.

Thanks,
Steven

**Comment by bstriner** [comment]
If your custom layer takes an input but doesn't actually use it in `call`, then it won't be added to the compute graph.

I mention getting batch_size from another input because you will typically have at least one input to your model, maybe to another layer. Just pass that input to your bias layer as well, but don't use it, or just use the shape from it.

Another avenue would be to absorb your bias layer into whatever the next layer is. Presumably, you are using the input at some point. You could create a larger layer, like BiasThenYaddaYadda, which takes an input, does your bias, then does whatever it does with the bias and the input.

If you're concerned about performance, a bias layer is a bad idea. You are intentionally creating outputs of shape (batch_size, k) that are redundant. Writing the bias into your next layer is going to be the best-performing option.

I've dealt with the checks in topology.py and you're better off having a slightly weird model definition than trying to fight it. That is something that would be worth enhancing but would be a long discussion.

Cheers,
Ben

**Comment by quantlearn** [comment]
I agree with you.   As with all things, there is an effort and efficiency trade off.  For now I have something working.  Thanks for your input. 

I am reasonably new to Keras and would like to understand the internals much better.   I love the functional API design but there are many custom layers that I need to create for some of our more esoteric models.   

I would be very interested in having a dialogue regarding ideas to enhance `topology.py`.   One of my initial observations is that shape inference can sometimes have a few hiccups.   I built a model with a TensorFlow backend that worked fine but the Keras backend complained about incompatible `ndims`.    I added a `Reshape` layer to make things explicit and then the code worked with both backends.   

I would like to see Keras extended to support a wider range of layers without sacrificing the core design principals.  Maybe it would be good to introduce a Tensorflow style contrib area.

Regards,
Steven

**Comment by bstriner** [comment]
Actually, a branch of Keras is being brought into tensorflow. Keras is also
working towards a keras_contrib repo. So it sounds like that will be ready
eventually but no idea on the timescale.

My wish for topology.py is better handling for things like autoencoders. I
only have one set of data I need to send to the GPU, but I always have to
send at least two arrays, even if one is a dummy.

The shape issues with TF and theano tend to relate to dim ordering. Theano
defaults to channels, w, h but TF is w,h, channels. If that was the issue,
doing a reshape will make things work but will actually be mixing the
channels and what you need is a Permute layer.

**Comment by quantlearn** [comment]
That is very interesting.  The keras_contrib repo could become a very powerful tool.

My personal areas of interest are Bayesian and Variational Inference methods in machine learning.  I am currently working on problems with high-dimensional time-series data and very large sparse datasets. 

Thank you for the tip regarding the Permute layer.  I will take a look at that.

Best regards,
Steven
